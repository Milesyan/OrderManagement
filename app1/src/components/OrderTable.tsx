import React, {useState} from 'react';
import Divider from '../common/Divider';
import {IOrder} from '../types';
import OrderTableCell, {OrderTableHeader} from './OrderTableCell';
import Search from './Search';

interface IOrderTable {
  orderData: IOrder[];
  supportSearch: boolean;
}
export default function OrderTable(props: IOrderTable) {
  const [searchTerm, setSearchTerm] = useState('');
  const _searchCents = Math.round(parseFloat(searchTerm) * 100);
  const shownOrderData = searchTerm ?
    props.orderData.filter(o => o.price === _searchCents)
    : props.orderData
  return (
    <>
      {
        props.supportSearch &&
        <>
          <Search onSearchTermUpdate={setSearchTerm} count={shownOrderData.length} />
          <Divider />
        </>
      }
      <table style={{width: '100%'}}>
        <thead>
          <OrderTableHeader />
        </thead>
        <tbody>
          {
            shownOrderData.map(o => (
              <OrderTableCell key={o.id} order={o} />
            ))
          }
        </tbody>
      </table>
    </>
  );
}

