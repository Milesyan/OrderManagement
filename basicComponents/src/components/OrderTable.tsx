import React, {MutableRefObject, useEffect, useRef, useState} from 'react';
import Divider from '../common/Divider';
import {IOrder} from '../types';
import OrderTableCell, {OrderTableHeader} from './OrderTableCell';
import Search from './Search';

interface IOrderTable {
  orderDataMap: Map<string, IOrder>;
  priceCache: Map<number, Set<string>>;
  supportSearch: boolean;
}
export default function OrderTable(props: IOrderTable) {

  const [searchTerm, setSearchTerm] = useState('');
  const [searchResults, setSearchResults] = useState<IOrder[]>([])
  const ref = useRef<HTMLDivElement>(null);
  const [showCount, setShowCount] = useState(30);
  const shownKeys = Array.from(props.orderDataMap.keys()).slice(0, showCount);
  const shownOrdersData = searchTerm ? searchResults : shownKeys.map(k => props.orderDataMap.get(k) as IOrder)
  useEffect(() => {
    const callback = (entries: IntersectionObserverEntry[]) => {
      entries.forEach((e) => {
        if (e.intersectionRatio > 0 && shownKeys.length >=30) {
          setShowCount(c => c+5);
        }
      });
    };
    const observer = new IntersectionObserver(callback, { rootMargin: "600px" });
    observer.observe(ref.current!);
    return () => {
      observer?.unobserve(ref.current!);
      observer?.disconnect();
    };
  }, [shownKeys, ref.current])
  
  useEffect(() => {
    const _searchCents = Math.round(parseFloat(searchTerm) * 100);
    const _ids = props.priceCache.get(_searchCents) ?? [];
    setSearchResults(Array.from(_ids).map(_id => props.orderDataMap.get(_id) as IOrder).filter(o => !!o))
  }, [searchTerm, props.priceCache])

  return (
    <>
      {
        props.supportSearch &&
        <>
          <Search onSearchTermUpdate={setSearchTerm} count={ searchTerm ? shownOrdersData.length: props.orderDataMap.size} />
          <Divider />
        </>
      }
      <table style={{width: '100%'}}>
        <thead>
          <OrderTableHeader />
        </thead>
        <tbody>
          {
            shownOrdersData.map(o => (
              <OrderTableCell key={o.id} order={o} />
            ))
          }
        </tbody>
      </table>
      <div ref={ref} style={{ height: 0, width: 0 }} />
    </>
  );
}

