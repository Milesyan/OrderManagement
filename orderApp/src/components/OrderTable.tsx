import React, { useEffect, useRef, useState} from 'react';
import OrderTableCell, {OrderTableHeader} from './OrderTableCell';
import Table from 'basicComponents/components/Table';
import {IOrder} from 'src/types';

const INITAL_SHOW_ORDERS_COUNT = 30;

export interface IOrderTable {
  orderDataMap: Map<string, IOrder>;
  priceCache: Map<number, Set<string>>;
  supportSearch: boolean;
}
export default function OrderTable(props: IOrderTable) {
  const [searchTerm, setSearchTerm] = useState('');
  const [searchResults, setSearchResults] = useState<IOrder[]>([])
  const ref = useRef<HTMLDivElement>(null);
  const [showCount, setShowCount] = useState(INITAL_SHOW_ORDERS_COUNT);
  const shownKeys = Array.from(props.orderDataMap.keys()).slice(0, showCount);
  const shownOrdersData = searchTerm ? searchResults : shownKeys.map(k => props.orderDataMap.get(k) as IOrder)
  useEffect(() => {
    const callback = (entries: IntersectionObserverEntry[]) => {
      entries.forEach((e) => {
        if (e.intersectionRatio > 0 && shownKeys.length >= INITAL_SHOW_ORDERS_COUNT) {
          setShowCount(c => c + 5);
        }
      });
    };
    const observer = new IntersectionObserver(callback, {rootMargin: "600px"});
    observer.observe(ref.current!);
    return () => {
      observer?.unobserve(ref.current!);
      observer?.disconnect();
    };
  }, [shownKeys, ref.current])

  useEffect(() => {
    const _searchCents = Math.round(parseFloat(searchTerm) * 100);
    const _ids = props.priceCache.get(_searchCents) ?? [];
    setSearchResults(Array.from(_ids).map(_id => props.orderDataMap.get(_id) as IOrder).filter(o => !!o))
  }, [searchTerm, props.priceCache])

  return (
    <>
      <Table
        supportSearch={true}
        searchPlaceholder={'Search by price (e.g. 34.07)'}
        text={`Matching Orders: ${searchTerm ? shownOrdersData.length : props.orderDataMap.size}`}
        renderHeader={() => (
          <OrderTableHeader />
        )}
        renderContent={() =>
          <>
            {
              shownOrdersData.map(o => (
                <OrderTableCell key={o.id} order={o} />
              ))
            }
          </>
        }
        onSearchUpdate={setSearchTerm}
      />
      <div ref={ref} style={{height: 0, width: 0}} />
    </>
  );
}

