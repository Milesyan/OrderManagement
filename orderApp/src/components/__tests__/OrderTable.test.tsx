import OrderTable, {IOrderTable} from '../OrderTable';
import TestRenderer from 'react-test-renderer';

jest.mock('basicComponents/components/Table', 
  () => { 
    return (props: any) => (
      <>
        {props.supportSearch && <div>Search Slot</div>}
        <table style={{width: '100%'}}>
        <thead>
          {props.renderHeader?.()}
        </thead>
        <tbody>
          {props.renderContent?.()}
        </tbody>
      </table>
      </>
    )
  },
  { virtual: true }
);

beforeAll(() =>{
  // @ts-ignore
  global.IntersectionObserver = class IntersectionObserver {
    constructor() {}
  
    disconnect() {
      return null;
    }
  
    observe() {
      return null;
    }
  
    takeRecords() {
      return null;
    }
  
    unobserve() {
      return null;
    }
  };
})
test('render OrderTable', () => {
  const props: IOrderTable = {
    supportSearch: true,
    "orderDataMap": new Map(Object.entries({
      "b11b8a69": {
        "customer": "Karina Lopez",
        "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
        "event_name": "DRIVER_RECEIVED",
        "id": "b11b8a69",
        "item": "Fish tacos",
        "price": 8700,
        "sent_at_second": 30,
      },
      "0b63508d": {
        "customer": "Heather Hughes",
        "destination": "7465 Willis Fork, Lake Gabrielle, CA 92816",
        "event_name": "CREATED",
        "id": "0b63508d",
        "item": "Berry smoothie",
        "price": 8029,
        "sent_at_second": 5,
      },
    })),
    "priceCache": new Map(
      [[8700, new Set(['b11b8a69'])],
      [8029, new Set(['0b63508d'])]]
    )
  }
  const ele = TestRenderer.create(<OrderTable
    {...props}
  />)
  expect(ele).toMatchSnapshot()
});

