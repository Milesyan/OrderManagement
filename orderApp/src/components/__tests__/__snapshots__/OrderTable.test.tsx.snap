// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render OrderTable 1`] = `
Array [
  <div>
    Search Slot
  </div>,
  <table
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <thead>
      <tr
        className="cellRoot headerCell"
      >
        <th
          className="item"
        >
          Item
        </th>
        <th
          className="customer"
        >
          Customer Name
        </th>
        <th
          className="price"
        >
          Price
        </th>
        <th
          className="destination"
        >
          Destination
        </th>
        <th
          className="status"
        >
          Order Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="cellRoot"
      >
        <th
          className="item"
        >
          Fish tacos
        </th>
        <th
          className="customer"
        >
          Karina Lopez
        </th>
        <th
          className="price"
        >
          $87.00
        </th>
        <th
          className="destination"
        >
          89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007
        </th>
        <th
          className="status"
        >
          Driver Received
        </th>
      </tr>
      <tr
        className="cellRoot"
      >
        <th
          className="item"
        >
          Berry smoothie
        </th>
        <th
          className="customer"
        >
          Heather Hughes
        </th>
        <th
          className="price"
        >
          $80.29
        </th>
        <th
          className="destination"
        >
          7465 Willis Fork, Lake Gabrielle, CA 92816
        </th>
        <th
          className="status"
        >
          Created
        </th>
      </tr>
    </tbody>
  </table>,
  <div
    style={
      Object {
        "height": 0,
        "width": 0,
      }
    }
  />,
]
`;
